modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/snapshot/programs/SnapshotProcessor.scala

```diff
@@ -0,0 +1,258 @@
+          setSnapshot
+            .map(
+              _ =>
+                RedownloadPerformed(
+                  globalSnapshot.height,
+                  globalSnapshot.ordinal,
+                  globalSnapshot.hash,
+                  globalSnapshot.proofsHash,
+                  toAdd.map(_._1.proofsHash),
+                  acceptedToRemove,
+                  obsoleteToRemove
+                )
+            )
+    }
+
+  private def checkAlignment(globalSnapshot: GlobalSnapshot): F[Alignment] =
+    for {
+      acceptedInMajority <- globalSnapshot.blocks.toList.traverse {
+        case BlockAsActiveTip(block, usageCount) =>
+          block.hashWithSignatureCheck.flatMap(_.liftTo[F]).map(b => b.proofsHash -> (b, usageCount))
+      }.map(_.toMap)
+
+      GlobalSnapshotTips(gsDeprecatedTips, gsRemainedActive) = globalSnapshot.tips
+
+      result <- lastGlobalSnapshotStorage.get.flatMap {
+        case Some(last) if last.ordinal.next == globalSnapshot.ordinal && last.height == globalSnapshot.height =>
+          blockStorage.getBlocksForMajorityReconciliation(last.height, globalSnapshot.height).flatMap {
+            case MajorityReconciliationData(_, _, _, _, acceptedAbove) =>
+              val onlyInMajority = acceptedInMajority -- acceptedAbove
+              val toMarkMajority = acceptedInMajority.view.filterKeys(acceptedAbove.contains).mapValues(_._2)
+              lazy val toAdd = onlyInMajority.values.toSet
+              lazy val toReset = acceptedAbove -- toMarkMajority.keySet
+//              val deprecatedTipsToRemove = deprecatedTips -- gsDeprecatedTips.map(_.block.hash)
+//              val deprecatedTipsToAdd = gsDeprecatedTips.filterNot(d => deprecatedTips.contains(d.block.hash))
+
+              if (onlyInMajority.isEmpty)
+                Applicative[F].pure[Alignment](AlignedAtNewOrdinal(toMarkMajority.toSet))
+              else
+                Applicative[F]
+                  .pure[Alignment](RedownloadNeeded(toAdd, toMarkMajority.toSet, Set.empty, Set.empty, toReset))
+          }
+
+        case Some(last)
+            if last.ordinal.next == globalSnapshot.ordinal && last.height.value < globalSnapshot.height.value =>
+          blockStorage.getBlocksForMajorityReconciliation(last.height, globalSnapshot.height).flatMap {
+            case MajorityReconciliationData(_, _, waitingInRange, acceptedInRange, acceptedAbove) =>
+              val acceptedLocally = acceptedInRange ++ acceptedAbove
+              val onlyInMajority = acceptedInMajority -- acceptedLocally
+              val toMarkMajority = acceptedInMajority.view.filterKeys(acceptedLocally.contains).mapValues(_._2)
+              val acceptedToRemove = acceptedInRange -- acceptedInMajority.keySet
+              lazy val toAdd = onlyInMajority.values.toSet
+              lazy val toReset = acceptedLocally -- toMarkMajority.keySet -- acceptedToRemove
+              val obsoleteToRemove = waitingInRange -- onlyInMajority.keySet
```

Comment:
I see some parts are repeated. can you extract both cases as separate fns? the whole for..comp will be easier to read and follow